adafruit_matrixportal.py

```diff
@@ -0,0 +1,898 @@
+#
+# SPDX-License-Identifier: Unlicense
+"""
+`adafruit_matrixportal`
+================================================================================
+
+Helper library for the Adafruit RGB Matrix Shield + Metro M4 Airlift Lite.
+
+* Author(s): Melissa LeBlanc-Williams
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+* `Adafruit Metro M4 Express AirLift <https://www.adafruit.com/product/4000>`_
+* `Adafruit RGB Matrix Shield <https://www.adafruit.com/product/2601>`_
+* `64x32 RGB LED Matrix <https://www.adafruit.com/product/2278>`_
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+
+* Adafruit's Bus Device library: https://github.com/adafruit/Adafruit_CircuitPython_BusDevice
+"""
+
+import os
+import time
+import gc
+import board
+import busio
+from digitalio import DigitalInOut
+import neopixel
+import terminalio
+from adafruit_esp32spi import adafruit_esp32spi, adafruit_esp32spi_wifimanager
+import adafruit_esp32spi.adafruit_esp32spi_socket as socket
+from adafruit_bitmap_font import bitmap_font
+import displayio
+from adafruit_display_text.label import Label
+from adafruit_io.adafruit_io import IO_HTTP, AdafruitIO_RequestError
+import adafruit_requests as requests
+import supervisor
+import rtc
+import rgbmatrix
+import framebufferio
+
+try:
+    from secrets import secrets
+except ImportError:
+    print(
+        """WiFi settings are kept in secrets.py, please add them there!
+the secrets dictionary must contain 'ssid' and 'password' at a minimum"""
+    )
+    raise
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_PyPortal.git"
+
+# pylint: disable=line-too-long
+# pylint: disable=too-many-lines
+# you'll need to pass in an io username, width, height, format (bit depth), io key, and then url!
+IMAGE_CONVERTER_SERVICE = "https://io.adafruit.com/api/v2/%s/integrations/image-formatter?x-aio-key=%s&width=%d&height=%d&output=BMP%d&url=%s"
+# you'll need to pass in an io username and key
+TIME_SERVICE = (
+    "https://io.adafruit.com/api/v2/%s/integrations/time/strftime?x-aio-key=%s"
+)
+# our strftime is %Y-%m-%d %H:%M:%S.%L %j %u %z %Z see http://strftime.net/ for decoding details
+# See https://apidock.com/ruby/DateTime/strftime for full options
+TIME_SERVICE_STRFTIME = (
+    "&fmt=%25Y-%25m-%25d+%25H%3A%25M%3A%25S.%25L+%25j+%25u+%25z+%25Z"
+)
+LOCALFILE = "local.txt"
+# pylint: enable=line-too-long
+
+
+class Fake_Requests:
+    """For faking 'requests' using a local file instead of the network."""
+
+    def __init__(self, filename):
+        self._filename = filename
+        with open(filename, "r") as file:
+            self.text = file.read()
+
+    def json(self):
+        """json parsed version for local requests."""
+        import json  # pylint: disable=import-outside-toplevel
+
+        return json.loads(self.text)
+
```

Comment:
Can we split this out into a separate library since it's shared with PyPortal? (I have modifications to this in my PyPortal copy and duplicating it here will leave it behind.)

Reply:
That's probably a good idea. There's a Blinka version of the PyPortal library as well that also shares this.